ceph/rbd/workflows/cluster_operations.py

```diff
@@ -0,0 +1,172 @@
+
+    Args:
+        client_node (Node): The ceph client node.
+
+    Raises:
+        Exception: If Ceph health check fails.
+    """
+    log.info("Checking Ceph health status")
+    try:
+        out, error = client_node.exec_command(cmd="ceph -s", sudo=True)
+
+        if "HEALTH_ERR" in out:
+            log.debug(out)
+            raise Exception("Cluster went to error health state")
+
+        log.info(f"Ceph health status: {out}")
+
+    except Exception as err:
+        log.error(err)
+        raise Exception("Failed to check Ceph health status")
+
+
+def osd_remove_and_add_back(
+    ceph_cluster,
+    rados_obj,
+    pool,
+):
+    """
+    Method to remove osd and add back the removed osd.
+
+    Args:
+        ceph_cluster: The Ceph cluster object.
+        rados_obj: RadosObject for interacting with Ceph.
+        pool: The name of the Ceph pool.
+
+    Raises:
+        Exception: If any ceph operation fails.
+    """
+    try:
+        client_node = ceph_cluster.get_nodes(role="client")[0]
+        pg_set = rados_obj.get_pg_acting_set(pool_name=pool)
+        log.info(f"Acting set for removal and addition of OSDs {pg_set}")
+        target_osd = pg_set[0]
+        host = rados_obj.fetch_host_node(daemon_type="osd", daemon_id=target_osd)
+
+        dev_path = get_device_path(host, target_osd)
+        log.info(
+            f"osd device path  : {dev_path}, osd_id : {target_osd}, host.hostname : {host.hostname}"
+        )
+
+        utils.set_osd_devices_unmanaged(ceph_cluster, target_osd, unmanaged=True)
+        method_should_succeed(utils.set_osd_out, ceph_cluster, target_osd)
+        method_should_succeed(wait_for_clean_pg_sets, rados_obj)
+        utils.osd_remove(ceph_cluster, target_osd)
+        method_should_succeed(wait_for_clean_pg_sets, rados_obj)
+        method_should_succeed(utils.zap_device, ceph_cluster, host.hostname, dev_path)
+        method_should_succeed(wait_for_device, host, target_osd, action="remove")
+
+        # Checking cluster health after OSD removal
+        method_should_succeed(rados_obj.run_pool_sanity_check)
+        log.info(
+            f"Removal of OSD : {target_osd} is successful. Proceeding to add back the OSD daemon."
+        )
+
+        # Adding the removed OSD back and checking the cluster status
+        utils.add_osd(ceph_cluster, host.hostname, dev_path, target_osd)
+
+        # Checking cluster health after OSD removal
+        method_should_succeed(rados_obj.run_pool_sanity_check)
+        cmd = f"ceph osd tree | grep {target_osd}"
+        timeout = 300
+        start_time = time.time()
+
+        while time.time() - start_time < timeout:
+            osd_tree_out = client_node.exec_command(cmd=cmd, output=True)[0]
+            if "up" in osd_tree_out:
+                log.info(
+                    f"Addition of OSD : {target_osd} back into the cluster is successful"
+                )
+                break
+
+            time.sleep(10)
+
+        else:
```

Comment:
I'm not sure if this else will work, what you should do is now since you are outside while loop, osd_tree_out will have "down" in it if the OSDs didn't come up. So check something like 
if "down" in osd_tree_out:
     raise Exception("")

Reply:
while `else` will take care that please refer here https://www.pythontutorial.net/python-basics/python-while-else/